homeassistant/components/airvisual/sensor.py

```diff
@@ -231,43 +196,8 @@ def unit_of_measurement(self):
                 }
             )
 
-
-class AirVisualData:
-    """Define an object to hold sensor data."""
-
-    def __init__(self, client, **kwargs):
-        """Initialize."""
-        self._client = client
-        self.city = kwargs.get(CONF_CITY)
-        self.country = kwargs.get(CONF_COUNTRY)
-        self.latitude = kwargs.get(CONF_LATITUDE)
-        self.longitude = kwargs.get(CONF_LONGITUDE)
-        self.pollution_info = {}
-        self.show_on_map = kwargs.get(CONF_SHOW_ON_MAP)
-        self.state = kwargs.get(CONF_STATE)
-
-        self.async_update = Throttle(kwargs[CONF_SCAN_INTERVAL])(self._async_update)
-
-    async def _async_update(self):
-        """Update AirVisual data."""
-
-        try:
-            if self.city and self.state and self.country:
-                resp = await self._client.api.city(self.city, self.state, self.country)
-                self.longitude, self.latitude = resp["location"]["coordinates"]
-            else:
-                resp = await self._client.api.nearest_city(
-                    self.latitude, self.longitude
-                )
-
-            _LOGGER.debug("New data retrieved: %s", resp)
-
-            self.pollution_info = resp["current"]["pollution"]
-        except (KeyError, AirVisualError) as err:
-            if self.city and self.state and self.country:
-                location = (self.city, self.state, self.country)
-            else:
-                location = (self.latitude, self.longitude)
-
-            _LOGGER.error("Can't retrieve data for location: %s (%s)", location, err)
-            self.pollution_info = {}
+    async def async_will_remove_from_hass(self) -> None:
+        """Disconnect dispatcher listener when removed."""
+        for cancel in self._async_unsub_dispatcher_connects:
+            self._async_unsub_dispatcher_connects.remove(cancel)
+            cancel()
```

Comment:
Do you really want to change content of a list your iterating over? Isn't it better to call cancel and after the loop reset the list